zebra-chain/src/parallel/tree.rs

```diff
@@ -160,36 +156,96 @@ impl NoteCommitmentTrees {
     }
 
     /// Update the sapling note commitment tree.
+    #[allow(clippy::unwrap_in_result)]
     fn update_sapling_note_commitment_tree(
         mut sapling: Arc<sapling::tree::NoteCommitmentTree>,
-        sapling_note_commitments: Vec<sapling::tree::NoteCommitmentUpdate>,
-    ) -> Result<Arc<sapling::tree::NoteCommitmentTree>, NoteCommitmentTreeError> {
+        mut sapling_note_commitments: Vec<sapling::tree::NoteCommitmentUpdate>,
+    ) -> Result<
+        (
+            Arc<sapling::tree::NoteCommitmentTree>,
+            Option<(u16, sapling::tree::Node)>,
+        ),
+        NoteCommitmentTreeError,
+    > {
         let sapling_nct = Arc::make_mut(&mut sapling);
+        let subtree_index_and_notes_in_next_subtree =
+            sapling_nct.subtree_index_and_notes_in_next_subtree(&mut sapling_note_commitments);
 
         for sapling_note_commitment in sapling_note_commitments {
             sapling_nct.append(sapling_note_commitment)?;
         }
 
+        // It is impossible for blocks to contain more than one level 16 sapling root:
+        // > [NU5 onward] nSpendsSapling, nOutputsSapling, and nActionsOrchard MUST all be less than 2^16.
+        // <https://zips.z.cash/protocol/protocol.pdf#txnconsensus>
+        //
+        // Before NU5, this limit holds due to the minimum size of Sapling outputs (948 bytes)
+        // and the maximum size of a block:
+        // > The size of a block MUST be less than or equal to 2000000 bytes.
+        // <https://zips.z.cash/protocol/protocol.pdf#blockheader>
+        // <https://zips.z.cash/protocol/protocol.pdf#txnencoding>
+        let mut subtree_root = None;
+
+        if let Some((subtree_index, sapling_note_commitments)) =
```

Comment:
Can we do the simple loop here?

Reply:
Yep, simplified in https://github.com/ZcashFoundation/zebra/pull/7371/commits/b77e050fe11e49afb91b5a58b2975f8eb5fb42d3.